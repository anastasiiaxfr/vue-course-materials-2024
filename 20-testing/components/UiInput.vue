<script setup>
import { ref } from 'vue'

defineOptions({
  inheritAttrs: false,
})
const props = defineProps({
  modelValue: {
    type: String,
    required: true,
  },
  rounded: {
    type: Boolean,
    default: false,
  },
  class: {
    type: [String, Object, Array],
  },
  inputClasses: {
    type: [String, Object, Array],
  },
})
const emit = defineEmits(['update:modelValue'])
const slots = defineSlots()
const model = defineModel()

const inputElement = ref(null)

function focus() {
  inputElement.value?.focus()
}

function hasLeftIcon() {
  return !!slots['left-icon']
}

defineExpose({
  focus,
})
</script>

<template>
  <div
      class="input-group"
      :class="[$props.class, {
      'input-group--icon': hasLeftIcon(),
      'input-group--icon-left': hasLeftIcon(),
    }]"
  >
    <div v-if="slots['left-icon']" class="input-group__icon">
      <slot name="left-icon" />
    </div>

    <input ref="inputElement"
           class="form-control"
           :class="[inputClasses, {
        'form-control--rounded': rounded,
      }]"
           v-bind="$attrs"
           v-model="model"
    />
  </div>
</template>

<style scoped>
/* _input-group.css */
.form-control {
  padding: 12px 16px;
  height: 52px;
  border-radius: 8px;
  border: 2px solid var(--blue-light);
  font-family: 'Nunito', sans-serif;
  font-weight: 600;
  font-size: 20px;
  line-height: 28px;
  color: var(--body-color);
  transition: 0.2s border-color;
  background-color: var(--white);
  outline: none;
  box-shadow: none;
}

.form-control::placeholder {
  font-weight: 400;
  color: var(--blue-2);
}

.form-control:focus {
  border-color: var(--blue);
}

textarea.form-control {
  width: 100%;
  min-height: 211px;
}

.form-control.form-control--rounded {
  border-radius: 26px;
}

.form-control.form-control--sm.form-control--rounded {
  border-radius: 22px;
}

.form-control.form-control--sm {
  padding: 8px 16px;
  height: 44px;
  border-radius: 4px;
}

.input-group {
  position: relative;
}

.input-group .form-control {
  width: 100%;
}

.input-group.input-group--icon-left .form-control {
  padding-left: 50px;
}

.input-group.input-group--icon-right .form-control {
  padding-right: 50px;
}

.input-group.input-group--icon .input-group__icon {
  position: absolute;
  top: 50%;
  transform: translate(0, -50%);
}

.input-group.input-group--icon.input-group--icon-left .input-group__icon:first-child {
  left: 16px;
}

.input-group.input-group--icon.input-group--icon-right .input-group__icon:last-child {
  right: 16px;
}
</style>
