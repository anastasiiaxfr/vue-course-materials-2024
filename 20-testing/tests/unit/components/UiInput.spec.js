import { vi, describe, it, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import UiInput from '../../../components/UiInput.vue'

// describe создаёт группу тестов
describe('UiInput', () => {
  // it описывает один тест. Алиас: test
  it('should be defined', () => {
    // expect - проверка теста
    expect(UiInput).toBeDefined()
  })

  it('should render input', () => {
    // mount - монтирует компонент в специальную обёртку
    const wrapper = mount(UiInput)
    const input = wrapper.find('input')
    expect(input.exists()).toBeTruthy()
  })

  // Можно создавать подгруппы тестов
  describe('rounded', () => {
    // Эти тесты тестируют вёрстку, ориентируясь на конкретную имплементацию (конкретные классы).
    // Это не очень хорошо, но хороших альтернатив нет:
    // - snapshot тесты
    // - screenshot тесты
    // Оба варианта ориентируются на эталон, который
    // - легко заменяется, но не всегда по валидной причине (нет контроля)
    // - может не пройти по множеству причин, а не только проверяет "скруглённость"

    it('should have .form-control--rounded on input with rounded', () => {
      const wrapper = mount(UiInput, {
        props: {
          rounded: true,
        },
      })
      const input = wrapper.get('input')
      expect(input.classes('form-control--rounded')).toBeTruthy()
    })

    it('should have .form-control--rounded on input without rounded', () => {
      const wrapper = mount(UiInput, {
        props: {
          rounded: false,
        },
      })
      const input = wrapper.get('input')
      expect(input.classes('form-control--rounded')).toBeFalsy()
    })

    it('should add .form-control--rounded on input rounded toggle', async () => {
      const wrapper = mount(UiInput, {
        props: {
          rounded: false,
        },
      })
      const input = wrapper.get('input')

      await wrapper.setProps({ rounded: true })
      // await nextTick() - не нужен, так как setProps уже его возвращает

      expect(input.classes('form-control--rounded')).toBeTruthy()
    })
  })

  it('should render input with modelValue', () => {
    const TEST_VALUE = 'test'
    const wrapper = mount(UiInput, {
      props: {
        modelValue: TEST_VALUE,
      },
    })
    const input = wrapper.get('input')
    expect(input.element.value).toBe(TEST_VALUE)
  })

  it('should emit update:modelValue on input', async () => {
    const wrapper = mount(UiInput)
    const input = wrapper.get('input')
    // input.element.value = 'test'
    // input.trigger('input')
    // input.trigger('change')
    // setValue делает всё перечисленное сам
    await input.setValue('test')

    expect(wrapper.emitted('update:modelValue')).toBeTruthy()
    expect(wrapper.emitted('update:modelValue')).toHaveLength(1)
    expect(wrapper.emitted('update:modelValue')[0]).toEqual(['test'])
  })

  it('should have model', async () => {
    let modelValue = 'test'
    const wrapper = mount(UiInput, {
      props: {
        modelValue,
      },
      attrs: {
        'onUpdate:modelValue': $event => {
          modelValue = $event
          wrapper.setProps({ modelValue })
        },
      },
    })

    const input = wrapper.get('input')
    await input.setValue('test2')

    expect(modelValue).toBe('test2')
    expect(wrapper.props('modelValue')).toBe('test2')
  })

  it('should set change handler on input', async () => {
    const handler = vi.fn()
    const wrapper = mount(UiInput, {
      attrs: {
        onChange: $event => {
          handler($event.target.value)
        },
      },
    })
    const input = wrapper.get('input')
    await input.setValue('test')

    expect(handler).toHaveBeenCalled()
    expect(handler).toHaveBeenCalledWith('test')
  })

  it('should match snapshot', () => {
    const wrapper = mount(UiInput)
    expect(wrapper.html()).toMatchSnapshot()
  })
})
