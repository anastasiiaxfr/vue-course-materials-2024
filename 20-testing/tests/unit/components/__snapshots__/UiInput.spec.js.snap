// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UiInput > should match snapshot 1`] = `
"<div data-v-a744b81a="" class="input-group">
  <!--v-if--><input data-v-a744b81a="" class="form-control">
</div>"
`;
